machxo2/pack.cc

```diff
@@ -19,326 +18,1567 @@
+                        if (can_add_flipflop_to_macro(comb, ci)) {
+                            ci->params[id_SD] = std::string("1");
+                            rel_constr_cells(comb, ci, (Arch::BEL_FF - Arch::BEL_COMB));
+                            // Packed successfully
+                            ++pairs;
+                            continue;
+                        }
+                    } else {
+                        // LUT/COMB is not part of a macro, this is the easy case
+                        // Constrain FF and LUT together, no need to rewire
+                        ci->params[id_SD] = std::string("1");
+                        comb->constr_children.push_back(ci);
+                        ci->cluster = comb->name;
+                        comb->cluster = comb->name;
+                        ci->constr_x = 0;
+                        ci->constr_y = 0;
+                        ci->constr_z = (Arch::BEL_FF - Arch::BEL_COMB);
+                        ci->constr_abs_z = false;
+                        // Packed successfully
+                        ++pairs;
+                        continue;
+                    }
+                }
+                {
+                    // Didn't manage to pack it with a driving combinational cell
+                    // Rewire to use general routing
+                    ci->params[id_SD] = std::string("0");
+                    ci->renamePort(id_DI, id_M);
                 }
             }
-            if (!packed_dff) {
-                lut_to_lc(ctx, ci, packed.get(), true);
+        }
+        log_info("    %d FFs paired with LUTs.\n", pairs);
+    }
+
+    // Return true if an port is a top level port that provides its own IOBUF
+    bool is_top_port(PortRef &port) { return false; }
+
+    // Return true if a net only drives a top port
+    bool drives_top_port(NetInfo *net, PortRef &tp)
+    {
+        if (net == nullptr)
+            return false;
+        for (auto user : net->users) {
+            if (is_top_port(user)) {
+                if (net->users.entries() > 1)
+                    log_error("   port %s.%s must be connected to (and only to) a top level pin\n",
+                              user.cell->name.c_str(ctx), user.port.c_str(ctx));
+                tp = user;
+                return true;
             }
-            new_cells.push_back(std::move(packed));
         }
+        if (net->driver.cell != nullptr && is_top_port(net->driver)) {
+            if (net->users.entries() > 1)
+                log_error("   port %s.%s must be connected to (and only to) a top level pin\n",
+                          net->driver.cell->name.c_str(ctx), net->driver.port.c_str(ctx));
+            tp = net->driver;
+            return true;
+        }
+        return false;
     }
 
-    for (auto pcell : packed_cells) {
-        ctx->cells.erase(pcell);
+    // Pass to pack LUT5s into a newly created slice
+    void pack_lut5xs()
```

Comment:
have you tested/used this at all?

Reply:
Removed this for now, easy to restore, did not update yosys part anyway so it is not used for now.